File: spk/cli/_cmd_make_binary.py
# Copyright (c) 2021 Sony Pictures Imageworks, et al.
# SPDX-License-Identifier: Apache-2.0
# https://github.com/imageworks/spk

from typing import Any
import argparse
import os

import structlog

import spk
from . import _flags

_LOGGER = structlog.get_logger("spk.cli")


def register(
    sub_parsers: argparse._SubParsersAction, **parser_args: Any
) -> argparse.ArgumentParser:

    mkb_cmd = sub_parsers.add_parser(
        "make-binary",
        aliases=["mkbinary", "mkbin", "mkb"],
        help=_make_binary.__doc__,
        description=_make_binary.__doc__,
        **parser_args
    )
    mkb_cmd.add_argument(
        "--here",
        action="store_true",
        help=("Build from the current directory, instead of a source package"),
    )
    mkb_cmd.add_argument(
        "--interactive",
        "-i",
        action="store_true",
        default=False,
        help="Setup the build, but instead of running the build script start an interactive shell",
    )
    mkb_cmd.add_argument(
        "--env",
        "-e",
        action="store_true",
        help="Build the first variant of this package, and then immediately enter a shell environment with it",
    )
    mkb_cmd.add_argument(
        "packages",
        metavar="PKG|SPEC_FILE",
        nargs="+",
        help="The packages or yaml specification files to build",
    )
    _flags.add_repo_flags(mkb_cmd)
    _flags.add_option_flags(mkb_cmd)
    _flags.add_runtime_flags(mkb_cmd)
    mkb_cmd.set_defaults(func=_make_binary)
    return mkb_cmd


def _make_binary(args: argparse.Namespace) -> None:
    """Build a binary package from a spec file or source package."""

    runtime = _flags.ensure_active_runtime(args)

    options = _flags.get_options_from_flags(args)
    repos = _flags.get_repos_from_repo_flags(args).values()
    for package in args.packages:
        if os.path.isfile(package):
            spec = spk.api.read_spec_file(package)
            _LOGGER.info("saving spec file", pkg=spec.pkg)
            spk.save_spec(spec)
        else:
            spec = spk.load_spec(package)

        _LOGGER.info("building binary package", pkg=spec.pkg)
        built = set()
        for variant in spec.build.variants:

            if not args.no_host:
                opts = spk.api.host_options()
            else:
                opts = spk.api.OptionMap()

            opts.update(variant)
            opts.update(options)
            if opts.digest() in built:
                continue
            built.add(opts.digest())

            _LOGGER.info("building variant", variant=opts)
            builder = (
                spk.BinaryPackageBuilder.from_spec(spec)
                .with_options(opts)
                .with_repositories(repos)
            )
            if args.here:
                builder = builder.with_source(os.getcwd())
            builder.set_interactive(args.interactive)
            try:
                out = builder.build()
            except (ValueError, spk.SolverError):
                _LOGGER.error("build failed", variant=opts)
                if args.verbose:
                    graph = builder.get_solve_graph()
                    print(spk.io.format_solve_graph(graph, verbosity=args.verbose))
                raise
            else:
                _LOGGER.info("created", pkg=out.pkg)
            if args.env:
                cmd = ["spk", "env", "-l", str(out.pkg)]
                _LOGGER.info("entering environment of new package", cmd=" ".join(cmd))
                os.execvp(cmd[0], cmd)

File: src/cli/cmd_make_binary.rs
// Copyright (c) 2022 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use std::sync::Arc;

use anyhow::{Context, Result};
use clap::Args;
use futures::TryFutureExt;
use spk::io::Format;
use spk::prelude::*;

use super::{flags, CommandArgs, Run};

#[derive(Clone)]
pub enum PackageSpecifier {
    Plain(String),
    WithSourceIdent((String, spk::api::RangeIdent)),
}

impl PackageSpecifier {
    // Return the package spec or filename string.
    fn get_specifier(&self) -> &String {
        match self {
            PackageSpecifier::Plain(s) => s,
            PackageSpecifier::WithSourceIdent((s, _)) => s,
        }
    }
}

impl std::str::FromStr for PackageSpecifier {
    type Err = clap::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // On the command line, only `Plain` is possible.
        Ok(PackageSpecifier::Plain(s.to_owned()))
    }
}

/// Build a binary package from a spec file or source package.
#[derive(Args)]
#[clap(visible_aliases = &["mkbinary", "mkbin", "mkb"])]
pub struct MakeBinary {
    #[clap(flatten)]
    pub repos: flags::Repositories,
    #[clap(flatten)]
    pub options: flags::Options,
    #[clap(flatten)]
    pub runtime: flags::Runtime,

    #[clap(short, long, global = true, parse(from_occurrences))]
    pub verbose: u32,

    /// Build from the current directory, instead of a source package)
    #[clap(long)]
    pub here: bool,

    /// Setup the build, but instead of running the build script start an interactive shell
    #[clap(long, short)]
    pub interactive: bool,

    /// Build the first variant of this package, and then immediately enter a shell environment with it
    #[clap(long, short)]
    pub env: bool,

    /// The packages or yaml spec files to build
    #[clap(name = "PKG|SPEC_FILE")]
    pub packages: Vec<PackageSpecifier>,

    /// Build only the specified variant, by index, if defined
    #[clap(long)]
    pub variant: Option<usize>,

    #[clap(flatten)]
    pub formatter_settings: flags::DecisionFormatterSettings,
}

impl CommandArgs for MakeBinary {
    // The important positional args for a make-binary are the packages
    fn get_positional_args(&self) -> Vec<String> {
        self.packages
            .iter()
            .map(|ps| ps.get_specifier())
            .cloned()
            .collect()
    }
}

#[async_trait::async_trait]
impl Run for MakeBinary {
    async fn run(&mut self) -> Result<i32> {
        let options = self.options.get_options()?;
        #[rustfmt::skip]
        let (_runtime, local, repos) = tokio::try_join!(
            self.runtime.ensure_active_runtime(),
            spk::storage::local_repository().map_ok(spk::storage::RepositoryHandle::from).map_err(anyhow::Error::from),
            async { self.repos.get_repos_for_non_destructive_operation().await }
        )?;
        let repos = repos
            .into_iter()
            .map(|(_, r)| Arc::new(r))
            .collect::<Vec<_>>();

        let mut packages: Vec<_> = self.packages.iter().cloned().map(Some).collect();
        if packages.is_empty() {
            packages.push(None)
        }

        for package in packages {
            let template = match flags::find_package_template(
                &package.as_ref().map(|p| p.get_specifier().to_owned()),
            )? {
                flags::FindPackageTemplateResult::NotFound(name) => {
                    Arc::new(spk::api::SpecTemplate::from_file(name.as_ref())?)
                }
                res => {
                    let (_, template) = res.must_be_found();
                    template
                }
            };

            tracing::info!("rendering template for {}", template.name());
            let recipe = template.render(&options)?;
            let ident = recipe.ident();

            tracing::info!("saving package recipe for {}", ident.format_ident());
            local.force_publish_recipe(&recipe).await?;

            tracing::info!("building binary package(s) for {}", ident.format_ident());
            let mut built = std::collections::HashSet::new();

            let variants_to_build = match self.variant {
                Some(index) if index < recipe.default_variants().len() => {
                    recipe.default_variants().iter().skip(index).take(1)
                }
                Some(index) => {
                    anyhow::bail!(
                        "--variant {index} is out of range; {} variant(s) found in {}",
                        recipe.default_variants().len(),
                        recipe.ident().format_ident(),
                    );
                }
                None => recipe.default_variants().iter().skip(0).take(usize::MAX),
            };

            for variant in variants_to_build {
                let mut opts = if !self.options.no_host {
                    spk::api::host_options()?
                } else {
                    spk::api::OptionMap::default()
                };

                opts.extend(variant.clone());
                opts.extend(options.clone());
                let digest = opts.digest_str();
                if !built.insert(digest) {
                    continue;
                }

                tracing::info!("building variant {}", spk::io::format_options(&opts));

                // Always show the solution packages for the solves
                let mut fmt_builder = self.formatter_settings.get_formatter_builder(self.verbose);
                let src_formatter = fmt_builder
                    .with_solution(true)
                    .with_header("Src Resolver ")
                    .build();
                let build_formatter = fmt_builder
                    .with_solution(true)
                    .with_header("Build Resolver ")
                    .build();

                let mut builder = spk::build::BinaryPackageBuilder::from_recipe(recipe.clone());
                builder
                    .with_options(opts.clone())
                    .with_repositories(repos.iter().cloned())
                    .set_interactive(self.interactive)
                    .with_source_resolver(&src_formatter)
                    .with_build_resolver(&build_formatter);

                if self.here {
                    let here =
                        std::env::current_dir().context("Failed to get current directory")?;
                    builder.with_source(spk::build::BuildSource::LocalPath(here));
                } else if let Some(PackageSpecifier::WithSourceIdent((_, ref ident))) = package {
                    // Use the source package `Ident` if the caller supplied one.
                    builder.with_source(spk::build::BuildSource::SourcePackage(ident.clone()));
                }
                let out = match builder.build_and_publish(&local).await {
                    Err(err @ spk::Error::Solve(_))
                    | Err(err @ spk::Error::PackageNotFoundError(_)) => {
                        tracing::error!("variant failed {}", spk::io::format_options(&opts));
                        return Err(err.into());
                    }
                    Ok((spec, _cmpts)) => spec,
                    Err(err) => return Err(err.into()),
                };
                tracing::info!("created {}", out.ident().format_ident());

                if self.env {
                    let request = spk::api::PkgRequest::from_ident(
                        out.ident().clone(),
                        spk::api::RequestedBy::CommandLine,
                    );
                    let mut cmd = std::process::Command::new(crate::env::spk_exe());
                    cmd.args(&["env", "--enable-repo", "local"])
                        .arg(request.pkg.to_string());
                    tracing::info!("entering environment with new package...");
                    tracing::debug!("{:?}", cmd);
                    let status = cmd.status()?;
                    return Ok(status.code().unwrap_or(1));
                }
            }
        }
        Ok(0)
    }
}
